Add admin tool for moving annotations from one URL to another

robertknight opened the pull request:
Add an admin tool at `/admin/documents` that enables moving all annotations on a set of URLs to a new set of URLs. The user of this tool needs to provide a JSON document which contains the mappings between the old URLs and the new URLs and their metadata. These mappings would typically be generated with a script such as the one in https://github.com/hypothesis/vitalsource-url-migration/tree/main.

The implementation reuses several pieces of the code that handle annotation updates via `PATCH /api/annotations`, to make sure that the annotation and document tables are updated in the same way.

Screenshot (pardon the ugliness of our admin UI):

<img width="1162" alt="Document URL migration screenshot" src="https://user-images.githubusercontent.com/2458/212326741-4edf25ca-1b2f-40ec-a815-6774be13dece.png">

**Part of https://github.com/hypothesis/h/issues/7709**

---

**Example usage:**

Before doing this, you must have `h-periodic` running locally.

1. Visit http://localhost:3000/document/burns and make an annotation
2. Open http://localhost:3000/document/doyle in another tab
3. Go to http://localhost:5000/admin/documents. Enter this JSON payload in the text field and click "Update URLs":

```json
{
  "http://localhost:3000/document/burns": {
    "url": "http://localhost:3000/document/doyle"
  }
}
```
4. You should see a message in the admin UI indicating that a URL migration has started.
5. Check the h debug logs and there should be some messages from Celery, but no errors
6. **Wait for h-periodic to trigger the "sync_annotations" task. This may take up to 60s** (_Perhaps we should trigger this automatically from the migrations task_)
7. Refresh the page at http://localhost:3000/document/burns and the annotation should have disappeared
8. Go to the tab containing http://localhost:3000/document/doyle and you should see the update notification. When you click it, the annotation that was originally made on the Burns poem should have moved to this page.

**Schema validation:**

Try entering text that doesn't parse as JSON, JSON that doesn't match the above schema or a mapping with non-HTTP(S) URLs in the admin UI textbox. You should see an error when submitting. The restriction to HTTP URLs is to catch accidental mistakes and is something we could relax in future.

**Advanced usage:**

The payload can optionally contain:

- A `document` field with document metadata for the new URL, in case it does not already exist in Hypothesis
- A `selectors` field with a list of selectors to add to migrated annotations. This is specifically useful when migrating annotations from book pages/chapter URLs to whole-book URLs. In this case the information about which chapter/page an annotation is on is captured in selectors rather than the URL, after the migration

<details>
<summary>Example of using `document` and `selectors` in the VitalSource migration</summary>

```json
{
  "https://jigsaw.vitalsource.com/books/0077387627/pages/650063327/content": {
    "url": "https://bookshelf.vitalsource.com/reader/books/0077387627",
    "selectors": [
      {
        "type": "EPUBContentSelector",
        "url": "https://jigsaw.vitalsource.com/books/0077387627/pages/650063327/content",
        "cfi": "/51",
        "title": "Chapter 02"
      },
      {
        "type": "PageSelector",
        "index": 51,
        "label": "52"
      }
    ],
    "document": {
      "title": "Essentials of Marketing Research"
    }
  }
}
```

</details>

seanh commented on h/services/url_migration.py:
```diff
@@ -0,0 +1,117 @@
+                    update_timestamp=False,
+                )
+
+                # Create an event that is the same as the one used when the
+                # annotation is updated via the API. We send it to consumers
+                # manually rather than using `request.notify_after_commit`
+                # because that mechanism is not available if we're being called
+                # from a Celery task. Also event handlers in `h.subscribers`
+                # try to start a new transaction, which we don't want.
+                event = AnnotationEvent(self.request, ann.id, "update")
+
+                # Re-index annotation in Elasticsearch.
+                search_index = self.request.find_service(name="search_index")
+                search_index.handle_annotation_event(event)
```
I can see why the convoluted-ness of the code around `SearchIndexService`, `storage.py`, etc is confusing but I don't think you want to call `search_index.handle_annotation_event()` here.

The call to `storage.update_annotation()` above will add the annotations to the job queue (a.k.a "asynchronous reindexing") which means they will get re-indexed in Elasticsearch as soon as the periodic reindexing task can process those jobs. The job queue/asynchronous reindexing is the appropriate thing to use here: it's designed to be able to [handle large batches of annotations at once from admin pages](https://docs.google.com/presentation/d/1msC6H5wxGQUo80zUWIe1rmfAAoQtUqIlrLFdBCMD-O4/edit#slide=id.ga6bcfe4558_1_34).

`search_index.handle_annotation_event()` is part of h's [synchronous annotation indexing](https://docs.google.com/presentation/d/1msC6H5wxGQUo80zUWIe1rmfAAoQtUqIlrLFdBCMD-O4/edit#slide=id.gc6fa3c898_0_28) which is meant to be used by the annotation create/update APIs. This is **not** designed to be able to handle large batches being added at once. I don't think you want to call this here.

See also: notes on [how and why h combines synchronous and asynchronous reindexing](https://docs.google.com/presentation/d/1msC6H5wxGQUo80zUWIe1rmfAAoQtUqIlrLFdBCMD-O4/edit#slide=id.g9f11104581_0_114)

seanh commented on h/services/url_migration.py:
```diff
@@ -0,0 +1,117 @@
+                search_index = self.request.find_service(name="search_index")
+                search_index.handle_annotation_event(event)
+
+                # Send real-time update notifications for clients watching the
+                # annotation or the new URL.
+                try:
+                    realtime_event_data = {
+                        "action": event.action,
+                        "annotation_id": event.annotation_id,
+                    }
+                    self.request.realtime.publish_annotation(realtime_event_data)
+                except RealtimeMessageQueueError as err:
+                    log.warning(
+                        f"Failed to send realtime update after URL migration: {err}"
+                    )
```
Hmm :) I think you can probably get away with not bothering to send realtime events. Other admin pages don't.

robertknight replied:
Indeed this is not essential, it was just convenient when testing to see the changes show up in the client immediately.

seanh replied:
On second thoughts I'd actually recommend leaving this out of the production version. It's inessential as you say and it might even cause a problem to trigger the system to send a whole batch of potentially tens-of-thousands of real-time notifications at once, which is not something it normally does

seanh reviewed (approved):
The direction of this draft PR looks broadly good to me :+1: My main feedback is that I think you want to let `storage.update_annotation()` place the annotations on the Elasticsearch reindexing job queue for you (and you need to add a `tag` argument to `update_annotation()`) and you **don't** want to call `SearchIndexService`. See comments for details

seanh commented on h/services/url_migration.py:
```diff
@@ -0,0 +1,117 @@
+                # Update the annotation's `target_uri` and associated document,
+                # and create `Document*` entities for the new URL if they don't
+                # already exist.
+                h.storage.update_annotation(
+                    self.request,
+                    ann.id,
+                    ann_update_data,
+                    # Don't update "edited" timestamp on annotation cards.
+                    update_timestamp=False,
+                )
```
Oh, I think you also need to pass `force=True` to the `add_by_id()` call that `storage.update_annotation()` makes, which means you'll need to add an optional `force` argument to `update_annotation()` as well. I don't think the code is working currently:

* `storage.update_annotation()` will [add annotations to the job queue](https://github.com/hypothesis/h/blob/26250b0fbc09430c1aef4d5972ae711ae9987829/h/storage.py#L192-L194) with `force=False` (see the default value of `force=False` in [`Queue.add_by_id()`](https://github.com/hypothesis/h/blob/26250b0fbc09430c1aef4d5972ae711ae9987829/h/services/search_index/_queue.py#L92-L102))
* What `force=True` does is it [forces re-indexing of an annotation even if its Elasticsearch already appears to be up to date](https://github.com/hypothesis/h/blob/26250b0fbc09430c1aef4d5972ae711ae9987829/h/services/search_index/_queue.py#L177-L189)
* With `force=False` the periodic reindexing task will [only reindex the annotation if the current document in Elasticsearch is not "equal" to the annotation in Postgres](https://github.com/hypothesis/h/blob/26250b0fbc09430c1aef4d5972ae711ae9987829/h/services/search_index/_queue.py#L204-L219), otherwise it will simply mark the job in the DB as completed without doing anything. "Equal" is defined as [if the Elasticsearch document has the same `user` and `updated` as the Postgres row](https://github.com/hypothesis/h/blob/26250b0fbc09430c1aef4d5972ae711ae9987829/h/services/search_index/_queue.py#L280-L286). Since you're not updating `updated` your annotations will be seen as equal and will not be reindexed

I think the reason why your annotations **are** being reindexed in your testing is because your service is also calling `search_index.handle_annotation_event()` (synchronous reindexing) below. So you're reindexing the annotation twice: once synchronously and once asynchronously. And the asynchronous attempt isn't working.

You definitely want to use the job queue/asynchronous reindexing for this, not the synchronous reindexing. See my comments on the `search_index.handle_annotation_event()` call below. The job queue/asynchronous reindexing **guarantees** that your annotation will be updated in Elasticsearch (it will keep on trying until it successfully reads back the correctly updated data from Elasticsearch) whereas the synchronous reindexing is "fire and forget" and does not guarantee that the reindex actually worked. You definitely want the "slower but guaranteed" version for batch reindexes like this

robertknight replied:
Thanks for the info. I started making changes along these lines but got confused / tripped up by several things along the way:

1. At first, the reindexing tasks did not get processed, because I didn't have `h_periodic` running locally.
2. When `h_periodic` is running normally, it only runs `sync_annotations` once every minute, which is fine for the automatic reindexing but is quite confusing if you're trying to use the admin console to trigger reindexing manually.
3. As I hadn't run h_periodic in a long time (maybe not ever) locally, my local h DB had a large volume of expired entries in the job queue, and it looks like [nothing removes them](https://hypothes-is.slack.com/archives/C1MA4E9B9/p1673873630354619)

robertknight commented on h/schemas/annotation.py:
```diff
@@ -24,41 +24,54 @@ def _validate_wildcard_uri(node, value):
 
+DOCUMENT_SCHEMA = {
+    "type": "object",
+    "properties": {
+        "dc": {
+            "type": "object",
+            "properties": {
+                "identifier": {"type": "array", "items": {"type": "string"}}
+            },
+        },
+        "highwire": {
+            "type": "object",
+            "properties": {
+                "doi": {"type": "array", "items": {"type": "string"}},
+                "pdf_url": {"type": "array", "items": {"type": "string"}},
+            },
+        },
+        "link": {
+            "type": "array",
+            "items": {
+                "type": "object",
+                "properties": {
+                    "href": {"type": "string"},
+                    "type": {"type": "string"},
+                },
+                "required": ["href"],
+            },
+        },
+    },
+}
+
+
+SELECTOR_SCHEMA = {
+    "type": "array",
+    "items": {
+        "type": "object",
+        "properties": {"type": {"type": "string"}},
+        "required": ["type"],
+    },
+}
```
Yes, exactly.

seanh commented:
One more thing occurred to me - we should check how many DB queries this makes, it would be bad if the number of queries scales with the number of annotations being migrated

seanh commented:
> One more thing occurred to me - we should check how many DB queries this makes, it would be bad if the number of queries scales with the number of annotations being migrated

Conversation is bouncing between GitHub and Slack, but indeed I think this is generating n queries per annotation and we're probably going to need to write bulk annotation update SQLAlchemy code: https://hypothes-is.slack.com/archives/C4K6M7P5E/p1673889916590389?thread_ts=1673615794.134599&cid=C4K6M7P5E

robertknight commented:
This has now been updated. Summary of changes since the last update:

1. Sean's feedback about triggering reindexing has been implemented. `h.storage.update_annotation` now takes optional arguments to force reindexing and set a job tag
2. Updates are now chunked up into Celery tasks in two stages. The mapping of URLs that the h admin submits are divided up into batches of URLs via Celery's `Task.chunks` API. Each of those tasks calls `URLMigrationService.move_annotations_by_url` which:
   1. Finds all annotations on the URL
   2. Moves the first annotation on that URL, in order to ensure the document entries are present in the DB and avoid `ConcurrentUpdateError`s from subsequent annotation moves
   3. The remaining annotations are then moved in separate batches via new Celery tasks

robertknight commented on h/jinja_extensions/navbar_data_admin.py:
```diff
@@ -43,6 +43,12 @@ def navbar_data_admin(request):
     },
+    {
+        "id": "documents",
+        "permission": Permission.AdminPage.HIGH_RISK,
```
I marked this as "high risk" for now because it is a destructive action. In time I expect we might want to make this self-service for customer support, perhaps with some added guardrails.

robertknight commented on h/storage.py:
```diff
@@ -180,18 +202,20 @@ def update_annotation(request, id_, data):
         )
 
-    if document:
+    if document or uri_changed:
```
This is a fix for a long-standing bug where an annotation's `document_id` did not get updated if an API caller called `POST /api/annotations` to update an annotation's URI, but didn't provide any document metadata.

seanh replied:
Is this an unrelated change? Could it be moved to a separate PR?

robertknight replied:
This change is needed in case you try to use the URL migration tool and only provide a new URL with no document metadata, it also fixes a bug which has long existed but probably has only been run into by third-parties before now so we hadn't noticed.

robertknight commented on h/streamer/messages.py:
```diff
@@ -127,7 +127,7 @@ def handle_annotation_event(message, sockets, request, session):
     for socket in matching_sockets:
         # Don't send notifications back to the person who sent them
-        if message["src_client_id"] == socket.client_id:
+        if socket.client_id and message.get("src_client_id") == socket.client_id:
```
This was needed at one point when the URL migration task was triggering realtime updates. This is no longer strictly required, but I think it makes sense to allow web-socket notifications to be dispatched without an associated `src_client_id` (this is an identifier for the Hypothesis client instance that triggered the notification by making an API call).

seanh replied:
This seems to be an unrelated change. Maybe make it in a separate PR, so you can deploy and test it separately and in case you need to revert it?

robertknight replied:
OK, I'll take this change out of this PR.

robertknight commented on h/tasks/url_migration.py:
```diff
@@ -0,0 +1,14 @@
+@celery.task
+def move_annotations_by_url(old_url, new_url_info):
+    migration_svc = celery.request.find_service(name="url_migration")
+    migration_svc.move_annotations_by_url(old_url, new_url_info)
```
I have used the term "move annotations" for these tasks and method names in `URLMigrationService`, because "moving an annotation between URLs" is how support will likely think about the operation, though earlier drafts used "update annotation URLs" in some places.

robertknight commented on h/views/admin/documents.py:
```diff
@@ -0,0 +1,53 @@
+
+    def report_error(self, message):
+        self.request.session.flash(message, "error")
+        self.request.response.status_code = 400
```
There wasn't a clear preferred pattern for making admin pages return 4xx pages in the event of a validation error, and IIRC some existing pages didn't handle this correctly in my testing.

seanh replied:
This should be handled using [exception views](https://docs.pylonsproject.org/projects/pyramid/en/latest/narr/views.html#custom-exception-views) but I don't know if h's admin pages have those setup properly yet.

robertknight replied:
Some pages make use of exception views, but it looks like this hasn't been set up in a standardized way yet. I think that would probably best be left as a separate refactoring exercise.

seanh replied:
:+1:

seanh commented on h/storage.py:
```diff
@@ -180,18 +202,20 @@ def update_annotation(request, id_, data):
         name="search_index"
-    )._queue.add_by_id(annotation.id, tag="storage.update_annotation", schedule_in=60)
+    )._queue.add_by_id(
+        annotation.id, tag=reindex_tag, schedule_in=0 if reindex else 60, force=reindex
```
```suggestion
        annotation.id, tag=reindex_tag, force=not update_timestamp
```

I don't think you need the `schedule_in` change here. As explained above `schedule_in=0` won't cause the reindexing to happen immediately. It will still take up to 30s or longer if there are more than 400 other jobs ahead of it on the queue. The job queue never does anything immediately. You don't actually need the `60s` delay in your case but for code simplicity I think you may as well leave it in.

For `force` I think you can do `force=not update_timestamp` and so avoid the need for the `reindex` argument.

seanh reviewed (changes requested):
I started looking at this but ran out of steam. Posting my comments so far, and I'll finish reviewing it tomorrow

robertknight commented:
> Are there now other places in the code that can be changed to doing annotation.uuid instead of doing UUID(types.URLSafeUUID.url_safe_to_hex(annotation.id)) themselves?

Likely yes. I figured it would be best to leave that for a separate PR.

> Posting my comments so far, nothing major although I think the new reindex argument to storage.update_annotation() can be removed.

Thanks. I have removed the `reindex` option from `storage.update_annotation` as suggested and also the WebSocket change. I do think it would avoid admin-user confusion if we could make it so that the first batch of reindexing happened near-immediately after the form was submitted, but this can be work for a follow-up PR.

seanh commented on tests/h/services/url_migration_test.py:
```diff
@@ -0,0 +1,227 @@
+        update_annotation.assert_not_called()
+        move_annotations_task.delay.assert_not_called()
+
+    @pytest.fixture
+    def transform_document(self, patch):
+        return patch("h.services.url_migration.transform_document")
+
+    @pytest.fixture
+    def update_annotation(self, patch):
+        return patch("h.storage.update_annotation")
+
+    @pytest.fixture
+    def move_annotations_task(self, patch):
+        return patch("h.services.url_migration.move_annotations")
```
I'd make these three autouse fixtures just because we don't want any unit tests to accidentally be calling the real code:

```suggestion
    @pytest.fixture(autouse=True)
    def transform_document(self, patch):
        return patch("h.services.url_migration.transform_document")

    @pytest.fixture(autouse=True)
    def update_annotation(self, patch):
        return patch("h.storage.update_annotation")

    @pytest.fixture(autouse=True)
    def move_annotations_task(self, patch):
        return patch("h.services.url_migration.move_annotations")
```

seanh commented on h/views/admin/documents.py:
```diff
@@ -0,0 +1,53 @@
+    def get(self):
+        return {}
+
+    @view_config(
+        request_method="POST",
+        request_param="update_annotation_urls",
+        require_csrf=True,
+        renderer="h:templates/admin/documents.html.jinja2",
+    )
+    def update_annotation_urls(self):
+        try:
+            url_mappings_json = json.loads(self.request.params["url_mappings"])
+        except ValueError as err:
+            return self.report_error(f"Failed to parse URL mappings: {err}")
+
+        schema = URLMigrationSchema()
+
+        try:
+            url_mappings = schema.validate(url_mappings_json)
+        except ValidationError as err:
+            return self.report_error(f"Failed to validate URL mappings: {err}")
```
I'm guessing this is endemic throughout the admin pages, but the way this should really work is something like this:

```suggestion
    @view_config(
        request_method="POST",
        request_param="update_annotation_urls",
        require_csrf=True,
        renderer="h:templates/admin/documents.html.jinja2",
        schema=URLMigrationSchema,
    )
    def update_annotation_urls(self):
        # If we get here then the request body has already been validated.
        validated_json = self.request.validated_json
        
        ...
```

That is: the view function itself doesn't have to pull the `"url_mappings"` param out of the request, deal with the case of a missing param, call `URLMigrationSchema.validate(), or deal with the case of a validation error. That would all be done by a Pyramid view predicate (the `schema=URLMigrationSchema` argument to `@view_config()`) which would send the appropriate error response if necessary or call the view if the request is valid.

We'd probably have to implement a custom view predicate to make this work. But you can see how much shorter it'd make the code of all the admin pages views (and their unit tests) while also making error handling behaviour consistent.

seanh commented on h/services/url_migration.py:
```diff
@@ -0,0 +1,141 @@
+            .options(load_only(Annotation.id, Annotation.target_uri))
+        )
+
+        ann_ids = [ann.id for ann in annotations]
```
Since you only need the annotation IDs and not the full annotations, there might be a way to make the query more efficient by only retrieving the IDs in the first place. I can't remember if off the top of my head though

robertknight replied:
> Since you only need the annotation IDs and not the full annotations, there might be a way to make the query more efficient by only retrieving the IDs in the first place.

This is what the `load_only` option above is intended to do.

seanh reviewed (approved):
Lots of little suggestions but overall this looks shippable to me :+1:

This seems like a nice overall technique that we could use in future for other admin pages that do other batch operations on annotations.

For completeness: the other way we could have gone would be to use h's homegrown in-DB "job queue": https://github.com/hypothesis/h/blob/main/h/models/job.py

The admin page would add `"move-annotation"` jobs to the queue, one for each annotation to be moved. We'd then write a `move_annotation()` celery task to consume these jobs from the queue and have h-periodic call that task periodically.

In a way I think the home-grown job queue may be more appropriate for batch processing tasks like this: it's a system that was designed for batch processing, it can accept any number of jobs (as long as you can get them into the DB table one way or another) and then will chew through them at a fixed speed without breaking a sweat.

Whereas I don't think celery is _really_ meant for "batch processing" where you queue up a large number of low-urgency jobs all at once and then work through them over time. I expect celery will handle this just fine, but I'm not sure about if the volume of jobs was really large.

robertknight commented on h/models/annotation.py:
```diff
@@ -136,6 +137,16 @@ class Annotation(Base):
         uselist=True,
     )
 
+    @property
+    def uuid(self):
+        """
+        Return the UUID representation of the annotation's ID.
+
+        Annotation IDs are stored in the DB as a UUID but represented in the app
+        and API in a different format. This property returns the UUID version.
+        """
+        return UUID(types.URLSafeUUID.url_safe_to_hex(self.id))
```
I looked into this but didn't find any places where there was exactly this code. What code has tended to do is to log the URL-safe annotation IDs (ie. `annotation.id`), since that is what is conveniently available, and left it as an exercise for the reader to map those to UUIDs if they are trying to find the corresponding DB entries.

seanh commented on h/tasks/url_migration.py:
```diff
@@ -0,0 +1,14 @@
+from h.celery import celery
+
+
+@celery.task
```
Oh, I think you need a `rate_limit` here!

robertknight replied:
What value do you suggest?
